python_benchmarks/test_rope.py

```diff
@@ -0,0 +1,184 @@
+        q,
+        start_indices=[0, 0, 0, 0],
+        end_indices=[batch_size, num_heads, seq_len, features_per_head // 2],
+        strides=[1, 1, 1, 1],
+    )
+    q_image = fd.ops.slice(
+        q,
+        start_indices=[0, 0, 0, features_per_head // 2],
+        end_indices=[batch_size, num_heads, seq_len, features_per_head],
+        strides=[1, 1, 1, 1],
+    )
+
+    # nvFuser has problems generating negation for bfloat.
+    q_image = fd.ops.cast(q_image, dtype=DataType.Float)
+    q_image = -q_image
+    q_image = fd.ops.cast(q_image, dtype=DataType.BFloat16)
+
+    q_rotated = fd.ops.cat([q_image, q_real], dim=-1)
+
+    cos = fd.ops.broadcast_in_dim(
+        cos, shape=[1, 1, seq_len, features_per_head], broadcast_dims=[2, 3]
+    )
+    sin = fd.ops.broadcast_in_dim(
+        sin, shape=[1, 1, seq_len, features_per_head], broadcast_dims=[2, 3]
+    )
+
+    out = q * cos + q_rotated * sin
+    out = fd.ops.cast(out, DataType.BFloat16)
+    fd.add_output(out)
+
+
+# Idea from @nikitaved: we split and concatenate the embeddings instead of `q`.
+# The embeddings are constant that can be precomputed. So we pay the overhead
+# of split and concatenation only once. The actual forward pass is merely
+# elementwise+reduction surrounded by some meta ops.
+def rope_without_cat_fusion(
+    fd: FusionDefinition,
+    batch_size: int,  # B
+    seq_len: int,  # S
+    num_heads: int,  # H
+    features_per_head: int,  # F
+) -> None:
+    q = fd.define_tensor(
+        shape=[batch_size, seq_len, num_heads, features_per_head],
+        dtype=DataType.BFloat16,
+    )
+    # `cos_sin_matrix` is essentially a batch (of size S*F/2) of 2x2 matrices
+    # laid out in a special way to keep computation simple.
+    #
+    # Using the notations in Figure 1 in https://arxiv.org/pdf/2104.09864.pdf,
+    # cos_sin_matrix[0] contains the following:
+    #
+    #   cos(θ_1),   -sin(θ1)
+    #   cos(θ_2),   -sin(θ2)
+    #   ...
+    #   cos(θ_F/2), -sin(θ_F/2)
+    #   ------------------------
+    #   sin(θ_1),   cos(θ_1)
+    #   sin(θ_2),   cos(θ_2)
+    #   ...
+    #   sin(θ_F/2), cos(θ_F/2)
+    #
+    # cos_sin_matrix[i] is similar but each θ is multiplied by `i+1`.
+    cos_sin_matrix = fd.define_tensor(
+        shape=[seq_len, 2, features_per_head // 2, 2],
+        dtype=DataType.BFloat16,
+    )
+
+    q = fd.ops.reshape(
+        q, new_shape=[batch_size, seq_len, num_heads, 2, features_per_head // 2]
+    )
+    q = fd.ops.permute(q, dims=[0, 2, 1, 4, 3])
+    q = fd.ops.broadcast_in_dim(
+        q,
+        shape=[batch_size, num_heads, seq_len, 1, features_per_head // 2, 2],
+        broadcast_dims=[0, 1, 2, 4, 5],
+    )
+
+    cos_sin_matrix = fd.ops.broadcast_in_dim(
+        cos_sin_matrix,
+        shape=[batch_size, num_heads, seq_len, 2, features_per_head // 2, 2],
+        broadcast_dims=[2, 3, 4, 5],
+    )
+
+    out = fd.ops.sum(q * cos_sin_matrix, [-1])
+    out = fd.ops.cast(out, DataType.BFloat16)
```

Comment:
Is this cast necessary -- same as for `rope_with_cat` above?

Reply:
Yes. nvFuser does implicit type promotion. Without this, the output type would be Float.